### PR TITLE
Incorrect time range overlapping detection in api server

### PR DESCRIPTION
**Pull request by cdlliuy**
#155061669.

This PR is used to fix #155061669.

Also, enhance the unit test of policy validation.  Expands API server's  total unit test case from 353 to 895.

### CODE REVIEW

**Comment by cdlliuy** [comment]
For reviewers, all testcases in  api/test/unit/validation/validationMiddleware.test.js are refactored with mocha "dynamic test case generating" feature.  No need to view diff, please just check the committed code ..

File: api/lib/validation/attributeValidator.js [hunk 1]
```diff
@@ -20,6 +20,19 @@ var compareDateTimes = function(startDateTime,endDateTime) {
   return moment(startDateTime,'YYYY-MM-DDTHH:mm')
     .isSameOrAfter(moment(endDateTime,'YYYY-MM-DDTHH:mm'));
 }
+var normalizeStartDate = function(object) {
```
Comment (boyang9527) [inline api/lib/validation/attributeValidator.js]: why we do normalization, start_date and end_date are required fields, aren't they?

Comment (cdlliuy) [reply api/lib/validation/attributeValidator.js]: @boyang9527 , start_date and end_date are optional fields, so it is allowed to set "start_date" or "end_date" only, the function "normalizeStartDate" and "normalizeEndDate" facilitates  the time overlapping detection if "start_date" or "end_date" not set.

File: api/test/unit/validation/attributeValidator.test.js [hunk 2]
```diff
@@ -13,31 +13,56 @@ describe('Validating Policy JSON properties',function(){
...
+        it('Should fail to validate the policy as instance_min_count is greater than instance_max_count without schedule defined',function(){
+          fakePolicy.schedules = null
+          attributeValidator.validatePolicy(fakePolicy,function(result){
+            expect(result[0]).to.have.property('stack').and.equal('instance_min_count 10 is higher or equal to instance_max_count 5 in policy_json');
+            expect(result[0]).to.have.property('message').and.equal('instance_min_count and instance_max_count values are not compatible');
+            expect(result[0]).to.have.property('property').and.equal('instance_min_count');
+          });
+        });
+
+        it('Should fail to validate the policy as instance_min_count is greater than instance_max_count without schedule defined',function(){
```
Comment (qibobo) [inline api/test/unit/validation/attributeValidator.test.js]: without scaling_rules defined?

File: api/test/unit/validation/attributeValidator.test.js [hunk 3]
```diff
@@ -13,31 +13,56 @@ describe('Validating Policy JSON properties',function(){
...
+      it('Should fail to validate the policy as instance_min_count is greater than instance_max_count with schedule defined',function(){
+          attributeValidator.validatePolicy(fakePolicy,function(result){
+            expect(result[0]).to.have.property('stack').and.equal('instance_min_count 10 is higher or equal to instance_max_count 5 in policy_json');
+            expect(result[0]).to.have.property('message').and.equal('instance_min_count and instance_max_count values are not compatible');
+            expect(result[0]).to.have.property('property').and.equal('instance_min_count');
+          });
+        });
+
+        it('Should fail to validate the policy as instance_min_count is greater than instance_max_count without schedule defined',function(){
+          fakePolicy.schedules = null
```
Comment (qibobo) [inline api/test/unit/validation/attributeValidator.test.js]: do we need to test "undefined"?

Comment (cdlliuy) [reply api/test/unit/validation/attributeValidator.test.js]: it is a regression case. 
we did have validate error when schedules is undefined previously.
These case didn't change, just reformatted with "context"

Comment (cdlliuy) [reply api/test/unit/validation/attributeValidator.test.js]: change to "delete fakePolicy.schedules" to test the situation of "undefined"..

Also, add more testcase to validate the policy when properties are set to null.

**Comment by boyang9527** [comment]
The travis failed with "The log length has exceeded the limit of 4 MB (this usually means that the test suite is raising the same exception over and over). The job has been terminated"

I guess this is caused by the schema validation...

**Comment by qibobo** [comment]
LGTM

**Pull request merged**